mypy/renaming.py

```diff
@@ -392,3 +386,162 @@ def record_assignment(self, name: str, can_be_redefined: bool) -> bool:
         else:
             # Assigns to an existing variable.
             return False
+
+
+class LimitedVariableRenameVisitor(TraverserVisitor):
+    """Perform some limited variable renaming in with statements.
+
+    This allows reusing a variable in multiple with statements with
+    different types. For example, the two instances of 'x' can have
+    incompatible types:
+
+       with C() as x:
+           f(x)
+       with D() as x:
+           g(x)
+
+    The above code gets renamed conceptually into this (not valid Python!):
+
+       with C() as x':
+           f(x')
+       with D() as x:
+           g(x)
+
+    If there's a reference to a variable defined in 'with' outside the
+    statement, or if there's any trickiness around variable visibility
+    (e.g. function definitions), we give up and won't perform renaming.
+
+    The main use case is to allow binding both readable and writable
+    binary files into the same variable. These have different types:
+
+        with open(fnam, 'rb') as f: ...
+        with open(fnam, 'wb') as f: ...
+    """
+
+    def __init__(self) -> None:
+        # Short names of variables bound in with statements using "as"
+        # in a surrounding scope
+        self.bound_vars: List[str] = []
+        # Names that can't be safely renamed, per scope ('*' means that
+        # no names can be renamed)
+        self.skipped: List[Set[str]] = []
```

Comment:
Its kind of obvious from the types that its a stack, but not from the comment, I didn't look at the type closely so I missed this. Maybe adjust comment?